readme.txt
Examiner - An Advanced Inspection Addon
---------------------------------------
With Examiner, you can check other players gear, talents, achievements, honor and arena team details.

When you inspect someone, it will show a stat summery of all their equipped gear combined.
Please note that these values are from gear alone, and will not include bonuses from buffs, talents or normal base stats.

Each player you inspect can be cached, that way you can look them up later, even when they are not around.
The option to cache others is disabled by default, you will have to enable caching on the "Config" page.

You can bind a key to inspect your target, you can even bind a key to inspect whoever is under your mouse (stealth inspection).
To set this up, open the Key Bindings dialog and look under "Examiner".

Examiner uses about ~230 kb of memory with no players cached. But if caching is enabled, memory usage can grow quickly.

You can always find the latest release of this addon on WoWInterface.com.
Post your comments, questions and suggestions for this addon on the addons comment page.

Comparing Gear
--------------
To compare one person's gear with another, you can mark a target for compare, by right clicking on the "Stats" button to get the drop down menu to open,
this menu will have an entry called "Mark for Compare". You can also use the shortcut by holding down Shift when clicking the "Stats" button.
When marked for compare, the Stats page will no longer show the actual stats of your inspected person, but will instead show the difference in stats
compared to the person you marked for compare.

Cache Filtering
---------------
If your number of cached entries is getting quite high, you may want a way to filter out specific players.
First you have to open the filter dialog, do this by right clicking on the Cache button and pick "Filter...", or use the shortcut by holding down the Shift key and then click the Cache button.
The filter works a bit similar to the /who function in WoW, so you can for example set the filter to "1-79" and it will show all non level 80s.
To only show a specific class, you can use the filter "class:deathknight", and only Death Knights will be shown.
You can add as many filter matches as you want, so using the filter: "class:hunter race:nightelf guild:awesome" will show all Night Elf hunters from the guild <Awesome>.
Filter matches cannot have spaces, you must write them without, that is why "Death Knight" has to be written as "DeathKnight", you can even just use "deathk" it does not have to match the entire thing.
You can use the filter type "sex" to match males or females. Using "sex:2" will match males, "sex:3" will match females.
There are many filter types, here are some examples: "class", "race", "realm", "guild", "guildRank", "sex" and "zone".

Model Frame Controls
--------------------
Left Click + Move:		Rotates the Model
Right Click + Move:		Moves the Model
Mousewheel:				Zoom
Ctrl + Left Click:		Change Background
Ctrl + Right Click:		Toggle Background

Slash Commands
--------------
The slash command for Examiner is /examiner or just /ex.

Although you probably wont need to use any slash commands as almost everything is available from the UI,
there are a few things which can only be done through slash commands.

/ex inspect <unit> or /ex i 'unit'
This one will inspect the given unit (target, focus, party3 etc). If no unit token is given, it will inspect target and then player if no target exist.

/ex si <itemLink>
This command will scan just a single item and list its combined stats in the chat frame.
Holding down ALT while the mouse is over an item in Examiner will show a tooltip with the stats like this.

/ex compare <itemLink1> <itemLink2>
Compares two items and lists the stat differences.

/ex arena <rating>
Shows you the amount of arena points the given rating gives for 2v2, 3v3 and 5v5.

/ex scale <value>
Changes the scale of the Examiner window.

/ex clearcache
Clears the entire cache of Examiner.

Lacking Features, Ideas & Problems
----------------------------------
- Use GetItemStats("link"| ?) & GetItemStatDelta("link"|?, "link"|?)
- You may see a problem where a player you inspected from mouseover, or loaded from the cache will not appear to have any gems, this is not really an Examiner problem, but has to do with the game not filling in the information fast enough.
  This seemed to have been caused by some changes Blizzard did for the 3.1 patch. There are some workarounds that could be done to avoid this though, which I might be adding later.
- Lack of space for module buttons, there are seven already, one more and Examiner will most likely blow, hmmmm maybe not, but at least it's starting to look kinda bad. I however have no good ideas as to this could be solved.
- Bring back the ability, as a new module, to send an inspected player to another person with Examiner.
- OnInspect will hide the activePage even though the module can show data in OnCacheLoaded.
- Sometimes the blizzard achievement frame will throw bugs around, I just have to find a better way to surpress it when Examiner is requesting achievements.
- The Glyphs module can only shows the glyphs of the active talent spec.

Special Thanks
--------------
- Chester, the original author of SuperInspect, who gave me the idea to make this mod.
- Haldamir of Gorgonnash, for the German translation.
- omosiro, who made the Korean translation.
- g3gg0, changes to the German translation & author to one of the SuperInspect versions.
- Pettigrow of Sinstralis, for the translation to French.
- Siphony of EU-Onyxia, updated German translation.
- Dridzt, for some of the required changes in the WotLK to make it function properly again.
- quicksylver, for updating the English patterns to WotLK.
- Darigaaz of SoulFlayer(RU) for the Russian translation.
- Hoofik of AshenVale for the 3.0 patch update of the Russian translation.
- Maxfunkey (aka Carambha of EU-Lordearon), patch 3.0.2 update of the German and Frensh translation.
- Thorakon (aka Pulgrim of EU-Alexstrasza), for the patch 3.0.3 update of the German translation.
- chkid, for Korean translation update for the 3.1 patch.
- ericyen, for the Taiwan translation for the 3.1 release.
- All the people using Examiner!

changelog.txt
Examiner Changes | 2010 Aug 14
------------------------------
- Fixed a bug in most modules, that could cause an error when linking an item, talent or achievement into chat.
- Feats: Increased the width slightly of the drop down category selector.
- Cache: Added a "Show Alts" filter. With this enabled, only your alts will be shown in the cache. Great for checking what gear they have. You must reinspect yourself on your alts for this to work.
- Scanner: Set bonuses are now once again properly scanned. Not certain when this broke.
- Glyphs: This module has been removed. I had originally added this because I hoped Blizzard would add glyph inspection into the API, but that never happened. Maybe one day it will return.

Examiner Changes | 2010 Jun 27
------------------------------
- Fixed the "ChatFrameEditBox" issues.

Examiner Changes | 2009 Dec 16
------------------------------
- Changed the TOC version to 30300 for the 3.3 patch.
- Glyphs: Added a check for empty string returned as realm for units out of range. This will fix the error where it reported the "Player-" not being online.

Examiner Changes | 2009 Nov 28
------------------------------
- Localisation: Updated the Russian translation, thanks to Hoofik.

Examiner Changes | 2009 Nov 03
------------------------------
- ItemSlots: Added option to show item levels all the time, and not just when holding down the ALT key.
- Scanner: Armor Penetration rating is now showing correct according to the new patch 3.2.2 value.
- Scanner: Gems with armor penetration should now be detected.

Examiner Changes | 2009 Aug 13
------------------------------
- Scanner: Ratings for dodge, parry and resilience has been corrected with the new patch 3.2 values.

Examiner Changes | 2009 Aug 09
------------------------------
- Changed the TOC version to 30200 for the 3.2 patch.
- Replaced the renamed "GetDifficultyColor" with "GetQuestDifficultyColor".
- Core: The delayed gem scan workaround will now properly re-cache the items.
- Stats: With the extra item tooltip lines added to in the 3.2 patch, the gem color detection should now work again.
- Feats: It's again possible to click an achievement to see its previous and following ones.
- Glyphs: Fixed a problem when requesting glyphs from players on other servers.
- Scanner: Negative stats are now detected, such as the "-5 Stamina" seen on "Fletcher's Gloves".
- Scanner: Optimised the line scanning a bit.
- Localisation: Updated the translation for the Taiwan version of WoW, thanks again ericyen.
- Localisation: The Korean translation has been updated, thanks chkid.

Examiner Changes | 2009 May 24
------------------------------
- Core: Changed the delayed gem inspection workaround from 0.5 seconds to 0.6.
- Core: The InspectUnit() function now supports a secondary parameter, setting this to false, makes Examiner not open. This could be useful for a plugin module that for example would inspect all people in the raid.
- Core: Auto caching of a player is now properly done after the inspection is done. This allows modules to easily add extra data to be cached.
- Stats: Prismatic gems now counts towards all colors for the "Gem Matches".
- Stats: No longer gives an error when inspecting someone who is naked or who is wearing gray items.
- Stats: The Average Item Level is now added to a cached entry.
- Stats: Adds a new sort method to the cache module: "iLvlAverage".
- Stats: Added a temporary slash command "/ex fixcacheitemlevels" that will give all your old cache entries and average itemlevel value, making it possible to sort cache by average item level.
- ItemSlots: When you hold down the Alt key, the item level of each item is shown on the item slot button.
- ItemSlots: Fixed a rare bug that could show a questionmark icon instead of an empty slot.
- ItemSlots: Fixed a bug caused by not clearing inspect data before reloading the cached entry when clicking on an item shown as a questionmark.
- Feats: The filter will now search the achievement rewards for a match.
- Feats: This module can now cache the number of achievement points a player has.
- Feats: Adds a new sort method to the cache module: "achievementPoints".
- Cache: Filtering has been improved, you can now set a filter similar to the /who function, for example a filter set to "1-79" will show all non 80s. There is also a class filter, use "class:deathknight" to show all Death Knights. Read more in the "readme.txt" file.
- Cache: Added two more entries to the cache menu, "Delete Shown Entries" and "Delete Hidden Entries". Using the new filtering, this is a good way to trim down your cached entries.
- Cache: The sort methods has been reworked, in the way that other modules can add new ways to sort cached entries. For examiner, the "Stats" module will add an "iLvlAverage" sort method.
- PvP: The Arena Point Calculator will now properly calculate points using 1500 rating if the given rating is less than 1500.
- Core: Added a new function "Examiner:GetModuleFromToken(token)" for modules to use if they need to interact with eachother.

Examiner Changes | 2009 May 10
------------------------------
- [Hotfix] It should now be possible to inspect others from a unitframe's right click menu again.
- Major rewrite of Examiner. It is now build up from modules, this means it will be a lot easier to update, edit and maintain. It is even possible to add third party plugin modules that adds extra features to Examiner.
- New Module: Glyphs. With this module you can examine other people's glyphs, although it does require them to have Examiner installed as well.
- Added a workaround for the missing gems when inspecting, it will not always work, especially not when inspecting a mouseover unit.
- It is now a lot more clear when watching a cached entry, then entire Examiner frame will get a slight tint of yellow.
- The "Scan Unknown Items" option has been removed, as it was obsolete.

Modules/config.lua
local ex = Examiner;

-- Module
local mod = ex:CreateModule("Config");
mod:CreatePage(false,"Configurations");
mod:CreateButton("Config","Configurations","Examiner Settings");

-- Create Version String
local modName = ex:GetName();
local vers = mod.page:CreateFontString(nil,"ARTWORK","GameFontHighlightSmall");
vers:SetText(modName.." |cffffff80"..GetAddOnMetadata(modName,"Version"));
vers:SetPoint("BOTTOM",0,14);

-- Variables
local cfg;
local checkBtns = {};

--------------------------------------------------------------------------------------------------------
--                                           Config Scripts                                           --
--------------------------------------------------------------------------------------------------------

-- CheckBoxes: OnClick
local function ConfigCheckBox_OnClick(self,button)
	local var = ex.options[self.id].var;
	cfg[var] = (self:GetChecked() ~= nil);
	-- Special "makeMovable" handling -- Must not invoke the OnHide event!
	if (var == "makeMovable") then
		local onHide = ex:GetScript("OnHide");
		ex:SetScript("OnHide",nil);
		ex:SetMovable(cfg.makeMovable);
		if (cfg.makeMovable) then
			HideUIPanel(ex);
			ex:Show();
		else
			ex:Hide();
			ShowUIPanel(ex);
		end
		ex:SetScript("OnHide",onHide);
	end
	-- Post Change to Modules
	for index, mod in ipairs(ex.modules) do
		if (mod.OnConfigChanged) then
			mod:OnConfigChanged(var,cfg[var]);
		end
	end
end

-- CheckBoxes: OnEnter
local function ConfigCheckBox_OnEnter(self)
	local option = ex.options[self.id];
	GameTooltip:SetOwner(self,"ANCHOR_RIGHT");
	GameTooltip:AddLine(option.label,1,1,1);
	GameTooltip:AddLine(option.tip,nil,nil,nil,1);
	GameTooltip:Show();
end

--------------------------------------------------------------------------------------------------------
--                                           Module Scripts                                           --
--------------------------------------------------------------------------------------------------------

-- Menu Init
local function MenuInit(parent,list)
	list[1].text = "|cff00c0ffExaminer Core Caching";
	list[1].value = "Core";
	list[1].checked = cfg.caching.Core;
	list[1].tip = "Disable this to turn off caching completely in Examiner";
	for index, mod in ipairs(ex.modules) do
		if (mod.canCache) then
			local tbl = list[#list + 1];
			tbl.text = mod.token;
			tbl.value = mod.token;
			tbl.checked = cfg.caching[mod.token];
			tbl.tip = "Enable to turn on caching for this module";
		end
	end
end

-- Menu Select
local function MenuSelect(parent,entry)
	cfg.caching[entry.value] = not cfg.caching[entry.value];
end

-- OnInitialize
function mod:OnInitialize()
	cfg = Examiner_Config;
	-- DropDown
	local dropDown = AzDropDown.CreateDropDown(self.page,190,false,MenuInit,MenuSelect);
	dropDown:SetPoint("TOP",0,-40);
	dropDown.label:SetText("Enabled Module Caching...");
	-- Check Boxes
	for index, option in ipairs(ex.options) do
		local chk = CreateFrame("CheckButton",nil,mod.page);
		chk:SetWidth(24);
		chk:SetHeight(24);
		chk:SetScript("OnClick",ConfigCheckBox_OnClick);
		chk:SetScript("OnEnter",ConfigCheckBox_OnEnter);
		chk:SetScript("OnLeave",ex.HideGTT);

		chk:SetNormalTexture("Interface\\Buttons\\UI-CheckBox-Up");
		chk:SetPushedTexture("Interface\\Buttons\\UI-CheckBox-Down");
	 	chk:SetHighlightTexture("Interface\\Buttons\\UI-CheckBox-Highlight");
		chk:SetDisabledCheckedTexture("Interface\\Buttons\\UI-CheckBox-Check-Disabled");
		chk:SetCheckedTexture("Interface\\Buttons\\UI-CheckBox-Check");

		chk:SetChecked(cfg[option.var]);

		chk.text = chk:CreateFontString("ARTWORK",nil,"GameFontNormalSmall");
		chk.text:SetPoint("LEFT",chk,"RIGHT",0,1);
		chk.text:SetText(option.label);
		chk:SetHitRectInsets(0,chk.text:GetWidth() * -1,0,0);

		chk.id = index;

		if (index == 1) then
			chk:SetPoint("TOPLEFT",20,-70);
		else
			chk:SetPoint("TOP",checkBtns[index - 1],"BOTTOM");
		end

		checkBtns[index] = chk;
	end
end

Modules/cache.lua
local ex = Examiner;

-- Module
local mod = ex:CreateModule("Cache");
mod:CreatePage(false,"");
mod:CreateButton("Cache","Cached Players","Right Click for extended menu");

-- Variables
local cacheSortMethods = { "none", "name", "realm", "level", "guild", "race", "class", "time" };
local BUTTON_HEIGHT = 24;
local cfg, cache;
local ExCacheList = {};
local buttons = {};
mod.cacheSortMethods = cacheSortMethods;

-- Filtering
local levelPattern = "%s(%d*)(%-)(%d*)%s";
local numberPattern = "%s%d+%s";
local listPattern = "([^%s]+):([^%s]+)";
local filterList = {};

-- Custom Colors
local CLASS_COLORS = CUSTOM_CLASS_COLORS or RAID_CLASS_COLORS;

-- RaceCoords. For females, add 0.5 to "top" and "bottom".
-- Interface\\Glues\\CharacterCreate\\UI-CharacterCreate-Races
local EX_RaceCoord = {
	Human		= { left = 0/8, right = 1/8, top = 0/4, bottom = 1/4 },
	Dwarf		= { left = 1/8, right = 2/8, top = 0/4, bottom = 1/4 },
	Gnome		= { left = 2/8, right = 3/8, top = 0/4, bottom = 1/4 },
	NightElf	= { left = 3/8, right = 4/8, top = 0/4, bottom = 1/4 },
	Draenei		= { left = 4/8, right = 5/8, top = 0/4, bottom = 1/4 },
	Tauren		= { left = 0/8, right = 1/8, top = 1/4, bottom = 2/4 },
	Scourge		= { left = 1/8, right = 2/8, top = 1/4, bottom = 2/4 },
	Troll		= { left = 2/8, right = 3/8, top = 1/4, bottom = 2/4 },
	Orc			= { left = 3/8, right = 4/8, top = 1/4, bottom = 2/4 },
	BloodElf	= { left = 4/8, right = 5/8, top = 1/4, bottom = 2/4 },
};

-- Dialog Func
local CacheFilterOkayFunc = function(text) cfg.cacheFilter = text; mod:BuildCacheList(); end

-- Page: OnShow
mod.page:SetScript("OnShow",function(self) if (#ExCacheList == 0) then mod:BuildCacheList(); end end);

-- HOOK: Slash Command "clearcache"
local oldClearCacheFunc = ex.slashFuncs.clearcache;
ex.slashFuncs.clearcache = function(cmd) oldClearCacheFunc(cmd); mod:BuildCacheList(); end

--------------------------------------------------------------------------------------------------------
--                                           Module Scripts                                           --
--------------------------------------------------------------------------------------------------------

-- OnInitialize
function mod:OnInitialize()
	cfg = Examiner_Config;
	cache = Examiner_Cache;
	-- Defaults
	cfg.cacheSort = cfg.cacheSort or "class";
	cfg.cacheFilter = cfg.cacheFilter or "";
end

-- OnButtonClick
function mod:OnButtonClick(button)
	-- left
	if (button == "LeftButton") then
		if (IsShiftKeyDown()) then
			AzDialog:Show("Enter new cache filter...",cfg.cacheFilter,CacheFilterOkayFunc);
		end
	-- right
	elseif (IsShiftKeyDown()) then
		cfg.cacheFilter = "";
		cfg.cacheShowAlts = false;
		self:BuildCacheList();
	end
end

-- OnCache -- Clear the cache list if this page isn't shown, to invalidate it
function mod:OnCache(entry)
	if (self.index == cfg.activePage) then
		self:BuildCacheList();
	else
		wipe(ExCacheList);
	end
end

--------------------------------------------------------------------------------------------------------
--                                       Delete Cache Functions                                       --
--------------------------------------------------------------------------------------------------------

-- Counts number of entries in a hash table
local function GetTableEntries(t)
	local count = 0;
	for _ in next, t do
		count = (count + 1);
	end
	return count;
end

-- Is the given value in the table?
local function IndexOf(t,value)
	for index, v in ipairs(t) do
		if (v == value) then
			return index;
		end
	end
end

-- Deletes all entries
local function DeleteAllCache()
	wipe(cache);
	mod:BuildCacheList();
end

-- Deletes all the shown entries
local function DeleteShownCache()
	for index, entryName in ipairs(ExCacheList) do
		cache[entryName] = nil;
	end
	mod:BuildCacheList();
end

-- Deletes all the hidden entries
local function DeleteHiddenCache()
	for entryName in next, cache do
		if (not IndexOf(ExCacheList,entryName)) then
			cache[entryName] = nil;
		end
	end
	mod:BuildCacheList();
end

--------------------------------------------------------------------------------------------------------
--                                                Menu                                                --
--------------------------------------------------------------------------------------------------------

-- Menu Init Items
function mod.MenuInit(parent,list)
	if (mod.index ~= cfg.activePage) then
		mod:BuildCacheList();
	end
	-- filter
	local tbl = list[#list + 1]; tbl.text = "Filter"; tbl.header = 1;
	tbl = list[#list + 1]; tbl.text = "Set Filter..."; tbl.value = 1;
	tbl = list[#list + 1]; tbl.text = "Show Alts"; tbl.value = 10; tbl.checked = cfg.cacheShowAlts;
	-- sort
	tbl = list[#list + 1]; tbl.header = 1;
	tbl = list[#list + 1]; tbl.text = "Sort Method"; tbl.header = 1;
	for index, method in ipairs(cacheSortMethods) do
		local tbl = list[#list + 1];
		tbl.text = method; tbl.value = method; tbl.checked = (cfg.cacheSort == method);
	end
	-- cache
	tbl = list[#list + 1]; tbl.header = 1;
	tbl = list[#list + 1]; tbl.text = "Cache"; tbl.header = 1;
	tbl = list[#list + 1]; tbl.text = "Delete All Entries"; tbl.value = 2;
	if (#ExCacheList ~= GetTableEntries(cache)) then
		tbl = list[#list + 1]; tbl.text = "Delete Shown Entries"; tbl.value = 3;
		tbl = list[#list + 1]; tbl.text = "Delete Hidden Entries"; tbl.value = 4;
	end
end

-- Menu Select Item
function mod.MenuSelect(parent,entry)
	-- Sort
	if (type(entry.value) == "string") then
		cfg.cacheSort = entry.value;
	-- Cache Filter
	elseif (entry.value == 1) then
		AzDialog:Show("Enter new cache filter...",cfg.cacheFilter,CacheFilterOkayFunc);
	elseif (entry.value == 10) then
		cfg.cacheShowAlts = (not cfg.cacheShowAlts);
	-- Delete Cache
	elseif (entry.value == 2) then
		AzDialog:Show("Are you sure you want to delete |cffffff80"..GetTableEntries(cache).."|r cached entries?",nil,DeleteAllCache);
	elseif (entry.value == 3) then
		AzDialog:Show("Sure you want to delete the |cffffff80"..#ExCacheList.."|r shown cache entries?",nil,DeleteShownCache);
	elseif (entry.value == 4) then
		AzDialog:Show("Sure you want to delete the |cffffff80"..(GetTableEntries(cache) - #ExCacheList).."|r hidden cache entries?",nil,DeleteHiddenCache);
	end
	-- Rebuild Cache List
	if (entry.value ~= 1) then
		mod:BuildCacheList();
	end
end

--------------------------------------------------------------------------------------------------------
--                                            Cache Stuff                                             --
--------------------------------------------------------------------------------------------------------

-- CacheEntry: OnClick
local function CacheEntry_OnClick(self,button)
	if (button == "LeftButton") then
		local editBox = ChatEdit_GetActiveWindow(); 
		if (editBox) and (IsModifiedClick("CHATLINK")) and (editBox:IsVisible()) then
			local entry = cache[self.entryName];
			editBox:Insert(format("%s, %d %s%s",self.entryName,entry.level,entry.class,(entry.guild and " of <"..entry.guild..">" or "")));
		else
			PlaySound("igMainMenuOptionCheckBoxOn");
			ex:ClearInspect();
			ex:LoadPlayerFromCache(self.entryName);
			-- Show Stats page if that module exists
			if (not IsShiftKeyDown()) then
				local mod = ex:GetModuleFromToken("Stats");
				if (mod) then
					ex:ShowModulePage(mod.index);
				end
			end
		end
	elseif (button == "RightButton") and (IsShiftKeyDown()) then
		cache[self.entryName] = nil;
		mod:BuildCacheList();
	end
end

-- CacheEntry: OnEnter
local function CacheEntry_OnEnter(self,motion)
	local entry = cache[self.entryName];
	GameTooltip:SetOwner(self,"ANCHOR_RIGHT");
	-- Init Text & Colors
	local color = CLASS_COLORS[entry.classFixed];
	local classText = ("|cff%.2x%.2x%.2x%s|r"):format(color.r*255,color.g*255,color.b*255,entry.class);
	color = GetQuestDifficultyColor(entry.level ~= -1 and entry.level or 500);
	local level = ("|cff%.2x%.2x%.2x%s|r"):format(color.r*255,color.g*255,color.b*255,entry.level ~= -1 and entry.level or "??");
	-- Add Lines & Show
	GameTooltip:AddLine(entry.pvpName..(entry.realm and " - "..entry.realm or ""),0.5,0.75,1.0);
	if (entry.guild) then
		GameTooltip:AddLine("<"..entry.guild..">",0.0,0.5,0.8);
	end
	GameTooltip:AddLine(level.." "..entry.race.." "..classText,1,1,1);
	GameTooltip:AddLine(entry.zone,1,1,1);
	GameTooltip:AddLine(ex:FormatTime(time() - entry.time));
	GameTooltip:AddLine(date("%a, %b %d, %Y - %H:%M:%S",entry.time));
	GameTooltip:AddLine("<Shift + Right-Click to Delete>",0.7,0.7,0.7);
	GameTooltip:Show();
end

-- Cache List Sorting
local function CacheListSortFunc(a,b)
	local aValue = cache[a][cfg.cacheSort];
	local bValue = cache[b][cfg.cacheSort];
	if (aValue == bValue) then
		return cache[a].name < cache[b].name;
	elseif (not aValue or not bValue) then
		return aValue and true;
	elseif (type(aValue) == "number") then
		return aValue > bValue;
	else
		return aValue < bValue;
	end
end

-- ScrollBar: Cache list update
local function UpdateShownItems()
	buttons[1]:SetWidth(#ExCacheList > #buttons and 200 or 216);
	FauxScrollFrame_Update(ExaminerCacheScroll,#ExCacheList,#buttons,BUTTON_HEIGHT);
	local gttOwner = GameTooltip:GetOwner();
	local index = ExaminerCacheScroll.offset;
	for i = 1, #buttons do
		index = (index + 1);
		local button = buttons[i];
		if (ExCacheList[index]) then
			local entryName = ExCacheList[index];
			local entry = cache[entryName];
			local color = CLASS_COLORS[entry.classFixed];

			button.entryName = entryName;
			button.name:SetFormattedText("|cffffffff%s|r %s",(entry.level ~= -1 and entry.level or "??"),entry.name);
			button.name:SetTextColor(color.r,color.g,color.b);

			button.value:SetText(cfg.cacheSort ~= "name" and entry[cfg.cacheSort]);

			local coords = EX_RaceCoord[entry.raceFixed];
			local iconOffset = (entry.sex == 3 and 0.5 or 0);
			button.race:SetTexCoord(coords.left,coords.right,coords.top + iconOffset,coords.bottom + iconOffset);

			-- Update Tooltip
			if (button == gttOwner) then
				CacheEntry_OnEnter(button);
			end

			button:Show();
		else
			button:Hide();
		end
	end
end

-- Include Filtered Entry
local function IncludeFilteredEntry(entry,filter,lvlMin,lvlHyphen,lvlMax)
	local include = true;
	-- Show Alt
	include = (include) and (not cfg.cacheShowAlts or cfg.cacheShowAlts and entry.isSelf);
	-- List Pattern
	for type, string in next, filterList do
		include = (include) and (entry[type]) and (tostring(entry[type]):lower():gsub(" ",""):match(string));
	end
	-- Level Match
	if (lvlMin and lvlMax) then
		include = (include) and (entry.level >= lvlMin and entry.level <= lvlMax);
	elseif (lvlHyphen == "-") then
		include = (include) and (lvlMin and entry.level >= lvlMin or lvlMax and entry.level <= lvlMax);
	elseif (lvlMin) then
		include = (include) and (entry.level == lvlMin);
	end
	-- Name Match & Return
	return (include) and (entry.name:lower():find(filter));
end

-- Build the table used to display the cache (display table)
function mod:BuildCacheList()
	-- We're padding the filter with spaces to fix some pattern matching
	local filter = (" "..cfg.cacheFilter.." ");
	-- Filter List
	for type, string in filter:gmatch(listPattern) do
		filterList[type] = string:lower();
	end
	-- Filter Level
	local lvlMin, lvlHyphen, lvlMax = filter:match(levelPattern);
	if (not lvlMin) then
		lvlMin = filter:match(numberPattern);
	end
	lvlMin, lvlMax = tonumber(lvlMin), tonumber(lvlMax);
	-- Filter Fin
	filter = filter:gsub(levelPattern,""):gsub(numberPattern,""):gsub(listPattern,""):trim():lower();
	local notFiltered = (filter == "" and not cfg.cacheShowAlts and not lvlMin and not lvlMax and not next(filterList));
	-- Create Display Table
	wipe(ExCacheList);
	for entryName, entry in next, cache do
		if (notFiltered) or (IncludeFilteredEntry(entry,filter,lvlMin,lvlHyphen,lvlMax)) then
			ExCacheList[#ExCacheList + 1] = entryName;
		end
	end
	-- Sort
	if (cfg.cacheSort ~= "none") then
		sort(ExCacheList,CacheListSortFunc);
	end
	-- Update
	self.page.header:SetFormattedText("Cached Players (%d)%s",#ExCacheList,(notFiltered and "" or " |cffffff00*"));
	UpdateShownItems();
	wipe(filterList);
end

--------------------------------------------------------------------------------------------------------
--                                        Cache Frame Creation                                        --
--------------------------------------------------------------------------------------------------------

-- Cache Entries
for i = 1, 10 do
	local btn = CreateFrame("Button",nil,mod.page);
	btn:SetWidth(200);
	btn:SetHeight(BUTTON_HEIGHT);
	btn:RegisterForClicks("LeftButtonDown","RightButtonDown");
	btn:SetHighlightTexture("Interface\\QuestFrame\\UI-QuestTitleHighlight");
	btn.id = i;

	if (i == 1) then
		btn:SetPoint("TOPLEFT",8,-40);
	else
		btn:SetPoint("TOPLEFT",buttons[i - 1],"BOTTOMLEFT");
		btn:SetPoint("TOPRIGHT",buttons[i - 1],"BOTTOMRIGHT");
	end

	btn:SetScript("OnClick",CacheEntry_OnClick);
	btn:SetScript("OnEnter",CacheEntry_OnEnter);
	btn:SetScript("OnLeave",ex.HideGTT);

	btn.race = btn:CreateTexture(nil,"ARTWORK");
	btn.race:SetTexture("Interface\\Glues\\CharacterCreate\\UI-CharacterCreate-Races");
	btn.race:SetPoint("LEFT",3,0);
	btn.race:SetWidth(BUTTON_HEIGHT - 2);
	btn.race:SetHeight(BUTTON_HEIGHT - 2);

	btn.name = btn:CreateFontString(nil,"ARTWORK","GameFontHighlight");
	btn.name:SetPoint("LEFT",btn.race,"RIGHT",3,0);
	btn.name:SetJustifyH("LEFT");

	btn.value = btn:CreateFontString(nil,"ARTWORK","GameFontHighlightExtraSmall");
	btn.value:SetPoint("LEFT",btn.name,"RIGHT",2,0);
	btn.value:SetPoint("RIGHT",-2);
	btn.value:SetJustifyH("RIGHT");

	buttons[i] = btn;
end

-- Cache Scroll
local scroll = CreateFrame("ScrollFrame","ExaminerCacheScroll",mod.page,"FauxScrollFrameTemplate");
scroll:SetPoint("TOPLEFT",buttons[1]);
scroll:SetPoint("BOTTOMRIGHT",buttons[#buttons],-3,-1);
scroll:SetScript("OnVerticalScroll",function(self,offset) FauxScrollFrame_OnVerticalScroll(self,offset,BUTTON_HEIGHT,UpdateShownItems) end);